src/looprestoration.rs

```diff
@@ -23,3 +27,188 @@ pub union LooprestorationParams {
     pub filter: Align16<[[int16_t; 8]; 2]>,
     pub sgr: LooprestorationParams_sgr,
 }
+
+macro_rules! looprestoration_filter_fn_enum {
+    (
+        pub enum $enum_name:ident;
+
+        {
+            $(
+                $rust_var:ident => { $rust_fn_16bpc:ident, $rust_fn_8bpc:ident },
+            )*
+        }
+
+        $(
+            #[$arch:meta]
+            {
+                $(
+                    $var_name:ident => { $fn_16bpc:ident, $fn_8bpc:ident },
+                )*
+            }
+        )*
+    ) => {
+        #[derive(Debug, Clone, Copy)]
+        #[repr(u8)]
+        pub enum $enum_name {
+            $(
+                $rust_var,
+            )*
+
+            $($(
+                #[cfg(all(feature = "asm", $arch))]
+                $var_name,
+            )*)*
+        }
+
+        impl $enum_name {
+            #[cfg(feature = "bitdepth_16")]
+            #[inline]
+            pub fn call_16bpc(
+                self,
+                dst: *mut pixel_16bpc,
+                dst_stride: ptrdiff_t,
+                left: const_left_pixel_row_16bpc,
+                lpf: *const pixel_16bpc,
+                w: libc::c_int,
+                h: libc::c_int,
+                params: *const LooprestorationParams,
+                edges: LrEdgeFlags,
+                bitdepth_max: libc::c_int,
+            ) {
+                match self {
+                    $(
+                        Self::$rust_var => unsafe {
```

Comment:
I don't think we want to mark these `unsafe` yet.  So far I've been leaving them `unsafe` until the `*_c`/`*_rust` versions are safe, so at least if we can show they're equivalent, it'd be safe (I'm not sure how `checkasm` works).